demos/common/src/images_capture.cpp

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "images_capture.h"
+
+#ifdef _WIN32
+#include "w_dirent.hpp"
+#else
+#include <dirent.h>
+#endif
+
+#include <opencv2/imgcodecs.hpp>
+
+#include <stdexcept>
+#include <string>
+#include <memory>
+
+class InvalidInput {};
+
+class ImreadWrapper : public ImagesCapture {
+public:
+    cv::Mat img;
+    bool canRead;
+
+    ImreadWrapper(const std::string &input, bool loop) : ImagesCapture{loop}, canRead{true} {
+        img = cv::imread(input);
+        if(!img.data) throw InvalidInput{};
+    }
+
+    double fps() const override {return 0.0;}
+
+    size_t lastImgId() const override {return 0;}
+
+    cv::Mat read() override {
+        if (loop) return img.clone();
+        if (canRead) {
+            canRead = false;
+            return img.clone();
+        }
+        return cv::Mat{};
+    }
+};
+
+class DirReader : public ImagesCapture {
+    std::vector<std::string> names;
+    size_t fileId;
+    size_t imgId;
+    const size_t posFrames;
+    const size_t readLengthLimit;
+
+public:
+    const std::string input;
+
+    DirReader(const std::string &input, bool loop, size_t posFrames, size_t readLengthLimit) : ImagesCapture{loop},
+            fileId{0}, imgId{0}, posFrames{posFrames}, readLengthLimit{readLengthLimit}, input{input} {
+        DIR *dir = opendir(input.c_str());
+        if (!dir) throw InvalidInput{};
+        while (struct dirent *ent = readdir(dir))
+            if (strcmp(ent->d_name, ".") && strcmp(ent->d_name, ".."))
+                names.emplace_back(ent->d_name);
+        closedir(dir);
+        if (names.empty()) throw InvalidInput{};
+        sort(names.begin(), names.end());
+        size_t readImgs = 0;
+        bool readAtLeastOnce = false;
+        while (fileId < names.size()) {
+            cv::Mat img = cv::imread(input + '/' + names[fileId]);
+            if (img.data) {
+                ++readImgs;
+                readAtLeastOnce = true;
+                if (readImgs - 1 >= posFrames) break;
```

Comment:
Isn't the `readAtLeastOnce` variable redundant? Comparing `readImgs` to `posFrames` already tells you whether the initialization is successful. You don't even need to repeat this check - just replace this `break` with `return` and then at the end of the loop you can throw unconditionally.

Reply:
Removed